src/Data/Singletons/Deriving/Show.hs

```diff
@@ -0,0 +1,140 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Data.Singletons.Deriving.Show
+-- Copyright   :  (C) 2015 Richard Eisenberg
+-- License     :  BSD-style (see LICENSE)
+-- Maintainer  :  Richard Eisenberg (rae@cs.brynmawr.edu)
+-- Stability   :  experimental
+-- Portability :  non-portable
+--
+-- Implements deriving of Show instances
+--
+----------------------------------------------------------------------------
+{-# LANGUAGE ScopedTypeVariables #-}
+module Data.Singletons.Deriving.Show (mkShowInstance) where
+
+import Language.Haskell.TH.Syntax hiding (showName)
+import Language.Haskell.TH.Ppr (pprint)
+import Language.Haskell.TH.Desugar
+import Data.Singletons.Names
+import Data.Singletons.Util
+import Data.Singletons.Syntax
+import Data.Singletons.Deriving.Infer
+import Control.Monad (when)
+import Data.Maybe (fromMaybe)
+import GHC.Lexeme (startsConSym, startsVarSym)
+import GHC.Show (appPrec, appPrec1)
+
+mkShowInstance :: DsMonad q => DType -> [DCon] -> q UInstDecl
+mkShowInstance ty cons = do
+  when (null cons) $
+    fail ("Can't derive Show instance for " ++ pprint (typeToTH ty) ++ ".")
+  clauses <- mk_showsPrec cons
+  return $ InstDecl { id_cxt = inferConstraints (DConPr showName) cons
+                    , id_name = showName
+                    , id_arg_tys = [ty]
+                    , id_meths = [ (showsPrecName, UFunction clauses) ] }
+
+mk_showsPrec :: DsMonad q => [DCon] -> q [DClause]
+mk_showsPrec cons = do
+    p <- newUniqueName "p" -- The precedence argument (not always used)
```

Comment:
In the code for `mk_showsPrec_clause`, I use `DWildPa`s in place of `DVarPa p` whenever `p` isn't used, so there's no need to prefix it with an underscore.